demos/TicTacToe/Classes/TicTacToeScene.cpp

```diff
@@ -1,26 +1,271 @@
+    if (snapshot.value().AsString() == wait_value_) {
+      got_value_ = true;
+    } else {
+      LogMessage(
+          "FAILURE: ExpectValueListener did not receive the expected result.");
+    }
+  }
+  void OnCancelled(const firebase::database::Error& error_code,
+                   const char* error_message) override {
+    LogMessage("ERROR: ExpectValueListener canceled: %d: %s", error_code,
+               error_message);
+  }
+
+  bool got_value() { return got_value_; }
+
+ private:
+  firebase::Variant wait_value_;
+  bool got_value_;
+};
+
+// Wait for a Future to be completed. If the Future returns an error, it will
+// be logged.
+void WaitForCompletion(const firebase::FutureBase& future, const char* name) {
+  while (future.status() == firebase::kFutureStatusPending) {
+    ProcessEvents(100);
+  }
+  if (future.status() != firebase::kFutureStatusComplete) {
+    LogMessage("ERROR: %s returned an invalid result.", name);
+  } else if (future.error() != 0) {
+    LogMessage("ERROR: %s returned error %d: %s", name, future.error(),
+               future.error_message());
+  }
+}
+char* random_string(std::size_t length) {
+  const std::string CHARACTERS = "0123456789abcdefghijkmnopqrstuvwxyz";
+
+  std::random_device random_device;
+  std::mt19937 generator(random_device());
+  std::uniform_int_distribution<> distribution(0, CHARACTERS.size() - 1);
+
+  char* random_string = new char[length + 1];
+
+  for (std::size_t i = 0; i < length; ++i) {
+    random_string[i] = CHARACTERS[distribution(generator)];
+  }
+  random_string[length] = '\0';
+
+  return random_string;
+}
+
+// A function that returns true if any of the row
+// is crossed with the same player's move
+bool rowCrossed(int board[][kTilesY]) {
+  for (int i = 0; i < kTilesY; i++) {
+    if (board[i][0] == board[i][1] && board[i][1] == board[i][2] &&
+        board[i][0] != kEmptyTile)
+      return (true);
+  }
+  return (false);
+}
+
+// A function that returns true if any of the column
+// is crossed with the same player's move
+bool columnCrossed(int board[][kTilesY]) {
+  for (int i = 0; i < kTilesX; i++) {
+    if (board[0][i] == board[1][i] && board[1][i] == board[2][i] &&
+        board[0][i] != kEmptyTile)
+      return (true);
+  }
+  return (false);
+}
+
+// A function that returns true if any of the diagonal
+// is crossed with the same player's move
+bool diagonalCrossed(int board[][kTilesY]) {
+  if (board[0][0] == board[1][1] && board[1][1] == board[2][2] &&
+      board[0][0] != kEmptyTile)
+    return (true);
+
+  if (board[0][2] == board[1][1] && board[1][1] == board[2][0] &&
+      board[0][2] != kEmptyTile)
+    return (true);
+
+  return (false);
+}
+
+// A function that returns true if the game is over
+// else it returns a false
+bool gameOver(int board[][kTilesY]) {
+  return (rowCrossed(board) || columnCrossed(board) || diagonalCrossed(board));
+}
+
+// Creating listeners for database values.
+SampleValueListener* current_player_index_listener;
```

Comment:
Is there a reason for these variables to be at the global scope? Could they belong to the TicTacToe class or something instead? Global variables are allowed when necessary, but generally you should avoid using them without a good reason.